File: monteCarloPi/monteCarloPiSeq.py
import sys
import time
import random

def calculate_pi(iterations):
    inside_circle = 0
    for _ in range(iterations):
        x = random.random()
        y = random.random()
        if x**2 + y**2 <= 1:
            inside_circle += 1
    return (4.0 * inside_circle) / iterations

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <iterations>")
        sys.exit(1)

    try:
        iterations = int(sys.argv[1])
        if iterations <= 0:
            raise ValueError
    except ValueError:
        print("Please provide a positive integer for iterations.")
        sys.exit(1)


    io_start_time = time.time()
    print(f"Starting Monte Carlo Pi computation with {iterations} iterations...")
    io_end_time = time.time()

    compute_start_time = time.time()
    pi = calculate_pi(iterations)
    compute_end_time = time.time()

    io_final_start = time.time()
    print(f"Estimated Pi: {pi:.10f}")
    io_final_end = time.time()

    io_time = (io_end_time - io_start_time) + (io_final_end - io_final_start)
    compute_time = compute_end_time - compute_start_time
    total_time = io_time + compute_time

    print(f"I/O Time: {io_time:.6f} seconds")
    print(f"Compute Time: {compute_time:.6f} seconds")
    print(f"Total Time: {total_time:.6f} seconds")

File: monteCarloPi/monteCarloPiSeq.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

double calculate_pi(long iterations) {
    long inside_circle = 0;
    for (long i = 0; i < iterations; i++) {
        double x = (double)rand() / RAND_MAX;
        double y = (double)rand() / RAND_MAX;
        if ((x * x + y * y) <= 1.0) {
            inside_circle++;
        }
    }
    return (4.0 * inside_circle) / iterations;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <iterations>\n", argv[0]);
        return 1;
    }

    long iterations = atol(argv[1]);
    if (iterations <= 0) {
        printf("Please provide a positive number of iterations.\n");
        return 1;
    }

    clock_t start_io = clock();
    printf("Starting Monte Carlo Pi computation with %ld iterations...\n", iterations);
    clock_t end_io = clock();

    clock_t start_compute = clock();
    double pi = calculate_pi(iterations);
    clock_t end_compute = clock();

    start_io = clock();
    printf("Estimated Pi: %.10f\n", pi);
    printf("I/O Time: %.6f seconds\n", ((double)(end_io - start_io) + (double)(clock() - end_io)) / CLOCKS_PER_SEC);
    printf("Compute Time: %.6f seconds\n", (double)(end_compute - start_compute) / CLOCKS_PER_SEC);
    printf("Total Time: %.6f seconds\n", (double)(clock() - start_io) / CLOCKS_PER_SEC);
    return 0;
}
